subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/GenericFileSystem.java

```diff
@@ -96,16 +97,22 @@ public GenericFileSystem(FileModeMutator chmod, FileModeAccessor stat, Symlink s
         this.symlink = symlink;
         this.chmod = chmod;
         canCreateSymbolicLink = symlink.isSymlinkSupported();
-        String content = generateUniqueContent();
-        File file = null;
-        try {
-            checkJavaIoTmpDirExists();
-            file = createFile(content);
-            caseSensitive = probeCaseSensitive(file, content);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
-        } finally {
-            FileUtils.deleteQuietly(file);
+    }
+
+    private void initializeCaseSensitive() {
+        if (caseSensitive == null) {
```

Comment:
I don't like this approach in general, as it's easy to overlook this when making changes. There should be _some_ kind of explicit declaration that this is and needs to be thread safe. Which could be a comment or an annotation or some code or something else.